module/os/linux/zfs/vdev_disk.c

```diff
@@ -1360,31 +1360,22 @@ vdev_disk_io_start(zio_t *zio)
 			return;
 		}
 
-		switch (zio->io_cmd) {
-		case DKIOCFLUSHWRITECACHE:
-
-			if (zfs_nocacheflush)
-				break;
-
-			if (v->vdev_nowritecache) {
-				zio->io_error = SET_ERROR(ENOTSUP);
-				break;
-			}
-
-			error = vdev_disk_io_flush(BDH_BDEV(vd->vd_bdh), zio);
-			if (error == 0) {
-				rw_exit(&vd->vd_lock);
-				return;
-			}
-
-			zio->io_error = error;
-
+		if (zfs_nocacheflush)
 			break;
```

Comment:
This path needs to somehow still call `rw_exit(&vd->vd_lock); zio_execute(zio); return;` not `break;`.  As it stands we also need to be careful not to fall-through to the end if the function which is marked `__builtin_unreachable()`.

Reply:
Oof, good spot - that's all on me, total halfarsery. Rewritten and I think good now.